redundant_closure fires for trait function that differs in levels of borrow from point-free form

`clippy 0.0.212 (d61b2541 2019-02-19)`

```rust
#![deny(clippy::redundant_closure)]

pub trait Dependency {
    fn required(&self) -> bool;
}

pub fn foo<D>(dependencies: &[D]) where D: Dependency {
    let _ = dependencies.iter().filter(|dep| dep.required());
}
```

```
error: redundant closure found
 --> src/lib.rs:8:40
  |
8 |     let _ = dependencies.iter().filter(|dep| dep.required());
  |                                        ^^^^^^^^^^^^^^^^^^^^ help: remove closure as shown: `Dependency::required`
```

But that suggestion will not compile because `dep` is a `&&D` and `Dependency::required` wants a `&D`

```rust
    let _ = dependencies.iter().filter(Dependency::required);
```

```
error[E0277]: the trait bound `&D: Dependency` is not satisfied
 --> src/lib.rs:9:40
  |
9 |     let _ = dependencies.iter().filter(Dependency::required);
  |                                        ^^^^^^^^^^^^^^^^^^^^ the trait `Dependency` is not implemented for `&D`
  |
note: required by `Dependency::required`
```

--- a/clippy_lints/src/eta_reduction.rs
+++ b/clippy_lints/src/eta_reduction.rs
@@ -133,23 +133,28 @@ fn get_ufcs_type_name(
     let actual_type_of_self = &cx.tables.node_type(self_arg.hir_id).sty;
 
     if let Some(trait_id) = cx.tcx.trait_of_item(method_def_id) {
-        //if the method expectes &self, ufcs requires explicit borrowing so closure can't be removed
-        return match (expected_type_of_self, actual_type_of_self) {
-            (ty::Ref(_, _, _), ty::Ref(_, _, _)) => Some(cx.tcx.item_path_str(trait_id)),
-            (l, r) => match (l, r) {
-                (ty::Ref(_, _, _), _) | (_, ty::Ref(_, _, _)) => None,
-                (_, _) => Some(cx.tcx.item_path_str(trait_id)),
-            },
-        };
+        if match_borrow_depth(expected_type_of_self, actual_type_of_self) {
+            return Some(cx.tcx.item_path_str(trait_id));
+        }
     }
 
     cx.tcx.impl_of_method(method_def_id).and_then(|_| {
-        //a type may implicitly implement other types methods (e.g. Deref)
+        //a type may implicitly implement other type's methods (e.g. Deref)
         if match_types(expected_type_of_self, actual_type_of_self) {
             return Some(get_type_name(cx, &actual_type_of_self));
         }
         None
     })
+}
+
+fn match_borrow_depth(lhs: &ty::TyKind<'_>, rhs: &ty::TyKind<'_>) -> bool {
+    match (lhs, rhs) {
+        (ty::Ref(_, t1, _), ty::Ref(_, t2, _)) => match_borrow_depth(&t1.sty, &t2.sty),
+        (l, r) => match (l, r) {
+            (ty::Ref(_, _, _), _) | (_, ty::Ref(_, _, _)) => false,
+            (_, _) => true,
+        },
+    }
 }
 
 fn match_types(lhs: &ty::TyKind<'_>, rhs: &ty::TyKind<'_>) -> bool {

--- a/tests/ui/eta.rs
+++ b/tests/ui/eta.rs
@@ -88,6 +88,14 @@ fn test_redundant_closures_containing_method_calls() {
     let c = Some(TestStruct { some_ref: &i })
         .as_ref()
         .map(|c| c.to_ascii_uppercase());
+
+    fn test_different_borrow_levels<T>(t: &[&T])
+    where
+        T: TestTrait,
+    {
+        t.iter().filter(|x| x.trait_foo_ref());
+        t.iter().map(|x| x.trait_foo_ref());
+    }
 }
 
 fn meta<F>(f: F)

